File: poc-csharp/RpsPoc/Domain/Match.cs
namespace RpsPoc.Domain;

public enum MatchState
{
    Pending,
    Running,
    Done,
}

/// <summary>
/// Partida entre dois jogadores. Escrita pela task do MatchRunner, lida pela
/// UI thread — os campos mutaveis sao publicados com volatile/Volatile.
/// </summary>
public sealed class Match(int id, Player p1, Player p2)
{
    private volatile MatchState _state = MatchState.Pending;
    private MatchResult? _result;
    private Round? _lastRound;

    public int Id { get; } = id;

    public Player P1 { get; } = p1;

    public Player P2 { get; } = p2;

    public MatchState State
    {
        get => _state;
        set => _state = value;
    }

    public MatchResult? Result
    {
        get => Volatile.Read(ref _result);
        set => Volatile.Write(ref _result, value);
    }

    public Round? LastRound
    {
        get => Volatile.Read(ref _lastRound);
        set => Volatile.Write(ref _lastRound, value);
    }
}

File: poc-java/src/main/java/rps/domain/Match.java
package rps.domain;

/**
 * Partida entre dois jogadores. Escrito pela thread do MatchRunner, lido pela
 * UI thread -> campos mutaveis sao volatile.
 */
public final class Match {

    public enum State { PENDING, RUNNING, DONE }

    private final int id;
    private final Player p1;
    private final Player p2;

    private volatile State state = State.PENDING;
    private volatile MatchResult result;
    private volatile Round lastRound;

    public Match(int id, Player p1, Player p2) {
        this.id = id;
        this.p1 = p1;
        this.p2 = p2;
    }

    public int id() {
        return id;
    }

    public Player p1() {
        return p1;
    }

    public Player p2() {
        return p2;
    }

    public State state() {
        return state;
    }

    public void state(State state) {
        this.state = state;
    }

    public MatchResult result() {
        return result;
    }

    public void result(MatchResult result) {
        this.result = result;
    }

    public Round lastRound() {
        return lastRound;
    }

    public void lastRound(Round round) {
        this.lastRound = round;
    }
}
